src/ts_catalog/dimension_partition.c

```diff
@@ -0,0 +1,414 @@
+		it,
+		Anum_dimension_partition_dimension_id_range_start_idx_dimension_id,
+		BTEqualStrategyNumber,
+		F_INT4EQ,
+		Int32GetDatum(dimension_id));
+	it->ctx.tuplock = tuplock;
+}
+
+/*
+ * Comparison function for dimension partitions.
+ *
+ * Used to sort arrays of partitions on range_start and to lookup a partition
+ * using binary search.
+ */
+static int
+dimpart_cmp(const void *key, const void *node)
+{
+	const DimensionPartition *dp_key = *((const DimensionPartition **) key);
+	const DimensionPartition *dp_node = *((const DimensionPartition **) node);
+
+	/* If the key's range is enclosed (or exactly matches) the partition,
+	 * there's a match */
+	if (dp_key->range_start >= dp_node->range_start && dp_key->range_end < dp_node->range_end)
+		return 0;
+	else if (dp_key->range_start < dp_node->range_start)
+		return -1;
+	else
+	{
+		Assert(dp_key->range_start > dp_node->range_start);
+		return 1;
+	}
+}
+
+static DimensionPartition *
+dimpart_create(const HeapTuple tup, const TupleDesc tupdesc)
+{
+	DimensionPartition *dp = palloc0(sizeof(DimensionPartition));
+	Datum values[Natts_dimension_partition];
+	bool isnull[Natts_dimension_partition] = { false };
+
+	heap_deform_tuple(tup, tupdesc, values, isnull);
+
+	dp->dimension_id =
+		DatumGetInt32(values[AttrNumberGetAttrOffset(Anum_dimension_partition_dimension_id)]);
+	dp->range_start =
+		DatumGetInt64(values[AttrNumberGetAttrOffset(Anum_dimension_partition_range_start)]);
+	dp->range_end = DIMENSION_SLICE_MAXVALUE;
+	dp->data_nodes = NIL;
+
+	if (!isnull[AttrNumberGetAttrOffset(Anum_dimension_partition_data_nodes)])
+	{
+		ArrayIterator arrit;
+		bool elem_isnull = false;
+		ArrayType *arr = DatumGetArrayTypeP(
+			values[AttrNumberGetAttrOffset(Anum_dimension_partition_data_nodes)]);
+		Datum elem = (Datum) NULL;
+
+		arrit = array_create_iterator(arr, 0, NULL);
+
+		while (array_iterate(arrit, &elem, &elem_isnull))
+		{
+			if (!elem_isnull)
+			{
+				const char *dn = NameStr(*DatumGetName(elem));
+				dp->data_nodes = lappend(dp->data_nodes, pstrdup(dn));
+			}
+		}
+
+		array_free_iterator(arrit);
+	}
+
+	return dp;
+}
+
+/*
+ * Get the current dimension partitions.
+ *
+ * The partitions are read into an array which is sorted so that it can be
+ * binary searched.
+ */
+DimensionPartitionInfo *
+ts_dimension_partition_info_get(int32 dimension_id)
+{
+	ScanIterator it;
+	DimensionPartitionInfo *dpi;
+	DimensionPartition **partitions;
+	unsigned int count = 0;
+#if TS_DEBUG
+	/* Use a low number of initial max partitions in DEBUG in order to "hit"
+	 * array expansion below */
+	unsigned int max_partitions = 2;
+#else
+	unsigned int max_partitions = 20;
+#endif
+
+	it = ts_dimension_partition_scan_iterator_create(AccessShareLock);
+	ts_dimension_partition_scan_iterator_set_dimension_id(&it, dimension_id, NULL);
+	/* Create a temporary array for partitions, which is likely too big */
+	partitions = palloc0(sizeof(DimensionPartition *) * max_partitions);
```

Comment:
nit: maybe there is a simple way to get a number of rows in the dimension partition table?